File: DMOJ/USACO/SocialDistancingI.py
import sys
from bisect import insort
input = sys.stdin.readline

n = int(input())
t = input()
cows = []


def check(x: int) -> bool:
    copy = cows
    cnt = 0
    if copy[0] >= x:
        insort(copy, 1)
        cnt += 1

    for i in range(len(copy) - 1):
        dis = copy[i+1] - copy[i]
        add = dis // (x+1) - 1
        cnt += (add if add >=0 else 0)
        if cnt >= 2:
            return True

    return False


for i in range(len(t)):
    if t[i] == '1':
        cows.append(i + 1)


lo = 1
hi = n
ans = 0
while lo <= hi:
    mid = (lo + hi) // 2
    if check(mid):
        ans = mid
        lo = mid + 1
    else:
        hi = mid - 1

print(ans + 1)

File: DMOJ/USACO/SocialDistancingI.cpp
// USACO 2020 Open Bronze P1 - Social Distancing I

#include <bits/stdc++.h>
using namespace std;

int n;
vector<int> cows;
bool check(int x) {
    vector<int> copy = cows;
    int cnt = 0;
}
int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    cin >> n;
    for (int i = 1; i <= n; i++) {
        char a;
        cin >> a;
        if (a == '1') cows.push_back(i);
    }

    int lo = 1, hi = n, ans = 0;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (check(mid)) {
            ans = mid;
            lo = mid + 1;
        } else
            hi = mid - 1;
    }

    cout << ans << '\n';

    return 0;
}